### PR TITLE
Initial commit of the approval resource

### PR DESCRIPTION
**Pull request by Seraf**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+# install requirements
+ADD requirements*.txt ./
+RUN pip install --no-cache-dir -r requirements.txt
+
+# install asserts
+ADD assets/ /opt/resource/
+
+# test
+ADD test/ /opt/resource-tests/
+#RUN /opt/resource-tests/test.sh
```
Comment (xlr-8) [inline Dockerfile:15]: Is the comment still necessary?

File: README.md [hunk 2]
```diff
@@ -1 +1,164 @@
...
+```
+# python3 cli.py approve --id f47f5864-a1b4-4bc0-8899-784ae3b768a4
+INFO: The lock f47f5864-a1b4-4bc0-8899-784ae3b768a4 has been approved
+```
+
+### Reject
+```
+# python3 cli.py reject --id f47f5864-a1b4-4bc0-8899-784ae3b768a4
+INFO: The lock f47f5864-a1b4-4bc0-8899-784ae3b768a4 has been rejected
+```
```
Comment (xlr-8) [inline README.md:167]: Nice README!
Could eventually add a license I guess?

File: ci/pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        params:
+          build: build-artifacts/
+
+  - name: test-unit
+    plan:
+      - get: resource-src
+        trigger: true
+        passed: [push-dev-image]
+      #- task: test-unit
+      #  file: resource-src/ci/tasks/test-unit.yml
```
Comment (xlr-8) [inline ci/pipeline.yml:18]: Leaving comments?
Couple more further down as well.

**Review by xlr-8 — COMMENTED** [review]
I have done a very brief review, it looks pretty good. I commented minor things, haven't gone deeply into the code.

File: assets/approval.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+            approval_lock = self.engine.query(Approval) \
+                .filter(
+                    pool=self.pool,
+                    lockname=params.get('lock_name')) \
+                .index('ts-index') \
+                .first(desc=True)
+
+            # We want to wait until the approve is done
+            if approval_lock.need_approval:
+                while approval_lock.approved is None:
```
Comment (gaelL) [inline assets/approval.py]: Maybe useless if statement.

We can use it directly on the while with both condition.

`while approval_lock.need_approval and approval_lock.approved is None`

Comment (gaelL) [reply assets/approval.py]: To clarify if statements maybe do something like : 

```
        while approval_lock.approved is None and approval_lock.need_approval:
            log.info("The lock %s is waiting for an approval" % params['lock_name'])
            # Query the lock item in the loop
            refresh_approval = self.query_lock(params['lock_name'])

            # If hasn't been approved or rejected, waiting a bit more
            if refresh_approval.approved is none:
                time.sleep(self.wait_lock)
                continue

            # Is it approved ?
            if refresh_approval.approved:
                approval_lock.approved = True
            # If not, we should fail the job and release the lock
            else:
                log.info("The lock hasn't been approved, exiting")
                approval_lock.claimed = False
                approval_lock.approved = None
                self.engine.save(approval_lock, overwrite=True)
                exit(1)
```

File: assets/approval.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+            # There is no approval, we have just a normal lock. Let's fetch the lock
+            approval_lock = self.engine.query(Approval)\
+                .filter(
+                    Approval.timestamp >= datetime.fromtimestamp(Decimal(version.get('timestamp'))),
+                    pool=self.pool)\
+                .index('ts-index')\
+                .first(desc=True)
+
+        metadata = []
+        if approval_lock:
```
Comment (gaelL) [inline assets/approval.py]: Replace directly the if by a if not to remove the else statement.
```
if not approval_lock:
    log.info("No lock have been found")
    exit(0)
```

File: assets/approval.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+            for key in approval_lock.keys_():
+                value = getattr(approval_lock, key)
+                if type(value) is datetime:
+                    value = str(Decimal(value.timestamp()))
+                if type(value) is bool:
```
Comment (gaelL) [inline assets/approval.py]: maybe `elif` to avoid to evaluate the statement again in case of the first one match

File: assets/approval.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+        if 'action' not in params:
+            log.error('You must set an action on params')
+
+        approval_lock = self.query_lock(params['lock_name'])
+        need_approval = params.get('need_approval', False)
+
+        if 'claim' in params['action']:
+            if approval_lock:
+                # If the lock is claimed in database
+                if approval_lock.claimed:
```
Comment (gaelL) [inline assets/approval.py]: Same as the comment on line 134

File: assets/approval.py [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+                log.info("Claiming the lock %s" % params['lock_name'])
+        elif 'release' in params['action']:
+            if approval_lock:
+                approval_lock.claimed = False
+                approval_lock.approved = None
+                approval_lock.timestamp = datetime.utcnow()
+                self.engine.save(approval_lock, overwrite=True)
+                log.info("Releasing the lock %s" % params['lock_name'])
+
+            else:
```
Comment (gaelL) [inline assets/approval.py]: Same as line 161

File: assets/approval.py [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+                log.info("The lock does not exist")
+                exit(1)
+        else:
+            log.error('Please use an available action')
+            exit(1)
+
+        metadata = []
+        for key in approval_lock.keys_():
+            value = getattr(approval_lock, key)
+            if type(value) is datetime:
```
Comment (gaelL) [inline assets/approval.py:320]: Same as line 165

File: assets/approval.py [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+
+        if 'lock_name' not in params:
+            log.error('You must set a lock_name on params')
+        if 'action' not in params:
+            log.error('You must set an action on params')
+
+        approval_lock = self.query_lock(params['lock_name'])
+        need_approval = params.get('need_approval', False)
+
+        if 'claim' in params['action']:
```
Comment (gaelL) [inline assets/approval.py:309]: WDYT about split actions in private functions like `_do_claim` and `_do_release` ?

File: assets/approval.py [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+        if self.command_name == 'check':
+            response = self.check_cmd(source, version)
+        elif self.command_name == 'in':
+            response = self.in_cmd(self.command_argument[0], source, version, params)
+        else:
+            response = self.out_cmd(self.command_argument[0], source, params)
+
+        return json.dumps(response)
+
+print(ApprovalResource(os.path.basename(__file__), sys.stdin.read(), sys.argv[1:]).run())
```
Comment (gaelL) [inline assets/approval.py]: Regarding python best practices for a script I would add `if __name__ == "__main__":` for this line.

And also used named arguments to be more readable like `info(spacing=10, collapse=1):`

File: ci/pipeline.yml (hunk 3 above)
Comment (Seraf) [reply ci/pipeline.yml:18]: Yes, I'm leaving them, I would like to add unit test in the future

File: README.md (hunk 2 above)
Comment (Seraf) [reply README.md:167]: Already done in https://github.com/cycloidio/approval-resource/blob/master/LICENSE

File: assets/approval.py (hunk 4 above)
Comment (Seraf) [reply assets/approval.py]: :+1:

File: assets/approval.py (hunk 5 above)
Comment (Seraf) [reply assets/approval.py]: :+1:

File: assets/approval.py (hunk 6 above)
Comment (Seraf) [reply assets/approval.py]: :+1:

**Pull request merged**